persist/fs/write.go

```diff
@@ -117,36 +120,71 @@ func NewWriter(opts Options) (FileSetWriter, error) {
 // Open initializes the internal state for writing to the given shard,
 // specifically creating the shard directory if it doesn't exist, and
 // opening / truncating files associated with that shard for writing.
-func (w *writer) Open(
-	namespace ident.ID,
-	blockSize time.Duration,
-	shard uint32,
-	blockStart time.Time,
-) error {
-	shardDir := ShardDirPath(w.filePathPrefix, namespace, shard)
-	if err := os.MkdirAll(shardDir, w.newDirectoryMode); err != nil {
-		return err
-	}
-	w.blockSize = blockSize
-	w.start = blockStart
+func (w *writer) Open(opts WriterOpenOptions) error {
+	var (
+		shardDir              string
+		fileTimestampUnixNano = opts.BlockStart
+		nextSnapshotIndex     int
+		err                   error
+	)
+
+	w.blockSize = opts.BlockSize
+	w.start = opts.BlockStart
+	w.snapshotTime = opts.SnapshotTime
 	w.currIdx = 0
 	w.currOffset = 0
-	w.checkpointFilePath = filesetPathFromTime(shardDir, blockStart, checkpointFileSuffix)
 	w.err = nil
 
 	var infoFd, indexFd, summariesFd, bloomFilterFd, dataFd, digestFd *os.File
-	if err := openFiles(
-		w.openWritable,
-		map[string]**os.File{
-			filesetPathFromTime(shardDir, blockStart, infoFileSuffix):        &infoFd,
-			filesetPathFromTime(shardDir, blockStart, indexFileSuffix):       &indexFd,
-			filesetPathFromTime(shardDir, blockStart, summariesFileSuffix):   &summariesFd,
-			filesetPathFromTime(shardDir, blockStart, bloomFilterFileSuffix): &bloomFilterFd,
-			filesetPathFromTime(shardDir, blockStart, dataFileSuffix):        &dataFd,
-			filesetPathFromTime(shardDir, blockStart, digestFileSuffix):      &digestFd,
-		},
-	); err != nil {
-		return err
+	switch opts.FilesetType {
+	case persist.FilesetSnapshotType:
+		shardDir = ShardSnapshotsDirPath(w.filePathPrefix, opts.Namespace, opts.Shard)
+		if err := os.MkdirAll(shardDir, w.newDirectoryMode); err != nil {
+			return err
+		}
+		nextSnapshotIndex, err = NextSnapshotFileIndex(w.filePathPrefix, opts.Namespace, opts.Shard, opts.BlockStart)
+		if err != nil {
+			return err
+		}
+		w.checkpointFilePath = snapshotPathFromTimeAndIndex(shardDir, fileTimestampUnixNano, checkpointFileSuffix, nextSnapshotIndex)
+
+		err := openFiles(
+			w.openWritable,
+			map[string]**os.File{
+				snapshotPathFromTimeAndIndex(shardDir, fileTimestampUnixNano, infoFileSuffix, nextSnapshotIndex):        &infoFd,
```

Comment:
any chance this code can be shared with the refactor you're going to do on the reader side?

Reply:
ended up not creating a separate struct so no, but I think the solution is fairly clean